aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -264,6 +268,40 @@ impl<T: Transaction> CapturedReads<T> {
         }
     }
 
+    pub(crate) fn capture_delayed_field_read(
+        &mut self,
+        id: T::Identifier,
+        value: AggregatorValue,
+    ) -> anyhow::Result<()> {
+        let ret = match self.delayed_field_reads.entry(id) {
+            Vacant(e) => {
+                e.insert(value);
+                UpdateResult::Inserted
+            },
+            Occupied(mut e) => {
+                let existing_read = e.get_mut();
+                if value == *existing_read {
+                    UpdateResult::Updated
+                } else {
+                    UpdateResult::Inconsistency(format!(
+                        "Read {:?} must be consistent with the already stored read {:?}",
+                        value, existing_read
+                    ))
+                }
+            },
+        };
+
+        match ret {
+            UpdateResult::IncorrectUse(_) => unreachable!(),
```

Comment:
setting self.incorrect_use = true and bailing seems like the proper behavior? bail makes sure the transaction is terminated fast (error returned from read), and irrespective of propagating proper error kind there, right after control flow comes back to us, we can (and should) check CapturedReads and the speculative error / code error flags and handle them.